test/operation-hooks.suite/dao-patch-by-id.suite.js

```diff
@@ -0,0 +1,227 @@
+// Copyright IBM Corp. 2015,2016. All Rights Reserved.
+// Node module: loopback-datasource-juggler
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+var ValidationError = require('../..').ValidationError;
+var contextTestHelpers = require('../helpers/context-test-helpers');
+var ContextRecorder = contextTestHelpers.ContextRecorder;
+var aCtxForModel = contextTestHelpers.aCtxForModel;
+var uid = require('../helpers/uid-generator');
+var HookMonitor = require('../helpers/hook-monitor');
+
+module.exports = function(dataSource, should, connectorCapabilities) {
+  describe('patchById', function() {
+    var instanceId, TestModel, hookMonitor, ctxRecorder, expectedError;
+    beforeEach(function setupHelpers() {
+      ctxRecorder = new ContextRecorder('hook not called');
+      hookMonitor = new HookMonitor({ includeModelName: true });
+      expectedError = new Error('test error');
+    });
+
+    beforeEach(function setupDatabase(done) {
+      TestModel = dataSource.createModel('TestModel', {
+        // Set id.generated to false to honor client side values
+        id: { type: String, id: true, generated: false, default: uid.next },
+        name: { type: String, required: true },
+        extra: { type: String, required: false },
+      });
+      done();
+    });
+
+    beforeEach(function setupData(done) {
+      TestModel.create({ name: 'John', extra: 'extra info' }, function(err, instance) {
+        if (err) return done(err);
+        instanceId = instance.id;
+        hookMonitor.install(TestModel);
+        done();
+      });
+    });
+
+    it('triggers hooks in the correct order', function(done) {
+      TestModel.patchById(instanceId,
+        { name: 'changed' },
+        function(err, info) {
+          if (err) return done(err);
+          hookMonitor.names.should.eql([
+            'TestModel:before save',
+            'TestModel:persist',
+            'TestModel:loaded',
+            'TestModel:after save',
+          ]);
+          done();
+        });
+    });
+
+    it('triggers `before save` hook', function(done) {
+      TestModel.observe('before save', ctxRecorder.recordAndNext());
+
+      TestModel.patchById(instanceId, { name: 'changed' }, function(err, info) {
+        if (err) return done(err);
+        ctxRecorder.records.should.eql(aCtxForModel(TestModel, {
+          where: { id: instanceId },
+          data: { name: 'changed' },
+        }));
+        done();
+      });
+    });
+
+    it('aborts when `before save` hook fails', function(done) {
+      TestModel.observe('before save', nextWithError(expectedError));
+
+      TestModel.patchById(instanceId, { name: 'changed' }, function(err, info) {
+        [err].should.eql([expectedError]);
+        done();
+      });
+    });
+
+    it('applies updates from `before save` hook', function(done) {
+      TestModel.observe('before save', function(ctx, next) {
+        ctx.data.extra = 'extra data';
```

Comment:
Isn't this the same value as you set when creating the model instance in a before-each hook above? I think the hook should not initialise the `extra` property.